### ISSUE TITLE
pic_id and file names which actually makes sense?

### ISSUE THREAD

**Issue by i92guboj** [jekkos/opensourcepos#1026]
Hi, this is neither a support request nor a feature request nor a bug report. I tried to locate a mailing list, which I think would be best for this kind of discussion but didn't find any. So, if this isn't the right place for this kind of topic, I apologize and kindly ask you to redirect me to whatever channel I should be using instead.

Right to the topic...

I have some concerns on how images are imported and stored. Right now, images can't be related to products in any meaningful way, which means you are bound to introduce them one by one. It seems ospos just puts the image in uploads/item_pics/, renames it as <number>.jpg, and puts <number> in ospos_items.pic_id

So, a nonsensical number gets stored in a field, and is arbitrarily used as a filename. I haven't looked at that part of the sources yet, I'll come to that later, tomorrow if I have the time. I've only been poking around with mysql-workbench...

Let's imagine, for a second, we changed pic_id to varchar. Let's imagine I have a lot of images for my products stored in such a typical way as <PRODUCT_REFERENCE>.png so that I can identify them quickly, as people normally do. And then, let's imagine I import my products using a CSV file, which, actually, I do.

All I'd need to add images for a thousand products would be to upload them into uploads/item_pics, and then run this SQL sentence:

`update ospos_items set pic_id=<field_containing_reference> `

It would be just too nice if I could also get a pic_id column in the default CSV template, that would even spare me the step of launching workbench.

Is this something that people would like or need or is it something I should be doing and keeping for myself? What would be the OSPOS-way to do it? 

And, again, forgive me if this isn't the right place for this. I really have no idea where else to ask.

Thanks!

**Comment by i92guboj** [comment]
I tried ALTERing that column, and in a sweep I was able to add pics to a good lot of products. It needs testing, it might break in some other place, but for now I can see the images for all my products in the product list.

**Comment by daN4cat** [comment]
Feel free to contribute for something more practical, this part can be changed and your help is welcomed.

**Comment by daN4cat** [comment]
I assume you will contribute for 3.1.

**Comment by jekkos** [comment]
I don't use this feature myself so I don't really care that much about it. If you have ideas to improve then feel free to propose some code changes in a PR to make it better. We will discuss and review before merging it upstream. Thanks.

**Comment by i92guboj** [comment]
Ok. Just wanted to know if this had any chance of getting upstream.

I will try to do something. I will have to think a bit about how to implement this, it would be a nice improvement albeit simple in terms of coding. Suggestions are welcome.

But I'll try to get something done as soon as I get some free time to work in this.

Thanks!

**Comment by i92guboj** [comment]
Just a quick question. Coding is one thing, but, at DB level, do you have any objections on changing the type of pic_id?

INT->VARCHAR(255) should not cause any issues, taking into account that that columns is not unique, nor it is a true ID of any kind (it's called pic_id, but other than that, it's NOT an ID).

I could use that as an effective ID and create a new table to host the actual file name, but that would be more complex, plus I never like the idea of creating a table consisting of ID+<other single field>.

Also, since it's NOT an ID, and you are not doing lookups based on that field, I don't think performance is an issue here. We could even change the field name to something more close to reality, such as "picture", "pic", or "pic_filename".

**Comment by daN4cat** [comment]
Change the field type and name. Feel free to do that, but don't create new tables for this.
With db change we need to take care of that in a db upgrade script and consider also how to migrate the current installations. Don't forget the backward compatibility.

**Comment by i92guboj** [comment]
Yes. I am considering that. But I don't plan to change the default behavior unless it's needed for some reason. All I want is to be able to use more reasonable patterns in that field (a name change is not strictly needed, it's just an aesthetic issue) and to be able to mass-import into that field.

The former, consist of an ALTER statement, and should be safe since it's INT to VARCHAR(255), no data loss should occur, though testing will be needed for sure.

The later is just a matter of changing the import process to include that field. 

This can be expanded in the future to add bells and all that. When we know the change is solid and doesn't randomly kill people's kittens. But that's secondary. 

For the rest, the way how images are uploaded and saved when you use the web UI doesn't need to be touched, and existing installations should just work. Again, testing is always a good thing.

With this in mind, if you feel there's something important I am missing, comments and guidance is always welcome. Otherwise, I'll just work in this as I manage to get some spare time. I don't know the ospos codebase so I'll take some grepping and trial to find my way, but it's small enough to be manageable which is something I like :)

**Comment by jekkos** [comment]
You mean in this case you will reference filenames from an excel and then upload them directly in the import function? In that case we should have some kind of generic upload method that can be called from within the import functionality. I chose to use number sequences at the time because file names need to be unique anyway. If we allow to use varchar in filenames then one can probably also choose to link one image to multiple items. Still I think that we should stick to using only one directory level on the server side.

**Comment by i92guboj** [comment]
I didn't think of auto uploading the files in the import process, though I suppose that could be done. I could put them in the folder myself. But if that's required to get this into upstream I will think about it.

As for linking images to more than one item, it can actually be done as the table is without any change. I did it before changing the column to varchar just to test something.

But yes. That's a must if you work with products of different sizes. It wouldn't make any sense to upload the same photo eleven times for the same shoe.

**Comment by i92guboj** [comment]
And thinking about that, maybe we should contemplate a way to "recycle" existing images when editing or creating products.

**Comment by jekkos** [comment]
Ok nvm about the excel import. If you think that it's useful then go ahead, but as I said I'm not really using it. 

I think the varchar type could be good indeed. And you're also correct on the name change, it might make more sense to name it pic_filename or the like.

Allowing file upload is a risky functionality as there are many ways to abuse even a slight misconfiguration and take over a PHP server with limited or elevated rights.

This is why I chose to use a very simple and straightforward naming scheme. If you choose to keep the name a user supplies then extra checks need to be done, which are eg. checking whether a .htaccess file is uploaded. If this is allowed then a user can reconfigure the serverside mimetype interpretation and eg. execute jpeg files as php scripts. There are numerous examples that embed php code inside JPEG's or PNG's metadata which can yield an easy server compromise.

Currently there are some content type filters active (those present in CI). More flexibility means more security risk in this case and I think we should consider every change carefully

**Comment by i92guboj** [comment]
I am no expert in PHP (or anything web-related) so just bear with me hehe.

Right now, with a couple of simple modifications, I am able to import my CSV file and fill my hand-tuned pic_id varchar field. It took me a bit to figure out a couple of things, since I am out of my league, but now I have a better ideal of the global picture.

Now, I need to figure out how to properly migrate existing databases. Converting the column is trivial, I just need to figure out how to fit it into ospos.

I'll keep you posted.

**Comment by i92guboj** [comment]
Hello, good morning!

This seems to be working, I've been using the functionality for a few days, so I thought it's time to take care of the db scripts. 

So far, I have created a 3.0_to_3.1.sql file, and put into it this:

> -- alter pic_id field, to contain a file name
> 
> ALTER TABLE `ospos_items` CHANGE `pic_id` `pic_filename` VARCHAR(255);

find -exec sed took care of changing the field name in all files under application/

I also modified tables.sql to set the same field name and type for that column. The rest of the files are supposed to be generated from that one, or so the comments at the top of those files say, but I don't know how.

So, my next question is, do I need to change some script for this to be applied when upgrading automatically or something? The sql is there, I don't know how to integrate this into the mechanics of ospos.

Please, forgive my noobness here, but I know little about web programming and nothing at all about this codeigniter thingy hehe.

**Comment by i92guboj** [comment]
Also, a bit off-topic, but would help me greatly with testing and debugging this: each time I need to re-improt the list, I first need to delete the contents from all the related tables. Otherwise, import will fail because of duplicates.

So, first question: you previously said you don't use the import functionality. Is there something I am missing or are you just creating and mass editing products by hand all the time? How are you supposed to make a 30% discount for everything, as an example... do you just use SQL as I do right now or is there something else?

Second, wouldn't it be desirable that, for existing items, they would be edited instead of inserted so A) you can actually change the items massively in a functional way and B) the import functionality don't fail because of such a silly thing?

**Comment by jekkos** [comment]
@i92guboj For SQL changes you indeed need to modify tables.sql and add the particular statement to the upgrade script. Next one needs to run a grunt build, see wiki for instructions on that. This will regenerate the final scripts.

The only way to edit items in bulk is by using bulk edit form. We could indeed update the import functionality to update instead of insert as well. 

But I just wonder which criteria one would use to match the lines in the import file with existing records. 

A) The only candidate is item_number which has to be unique. 
B) If this one is not filled then a row will be duplicated as it will always be considered as a new one? I don't think that this is something a user would want to achieve.
C) To solve this problem we can prompt the user if he's importing for the first time or merely updating. In latter case he needs to be aware of the possiblity that duplicate rows will be created in case there's no item_number specified.

**Comment by i92guboj** [comment]
Thanks for the pointers. I'll review that and report back when I have something working.

As for the import-to-edit part, I have to digest all that you said. It's for another topic anyway, but:
A) Agreed, it couldn't be any other, and it's a fine candidate anyway.
B) Is that field mandatory when importing? If not, it should. Is there any reason for it not to be mandatory?
C) I don't think that's necessary. it_exists?edit():create(); to sum it up

**Comment by jekkos** [comment]
OK just a short answer for the following

B) not sure whether it's mandatory in the import. I know it's certainly not in the database tables, as items can also be scanned using their Mysql PK. If the code does not enforce an item_number right now then it could break some functionality for other users out there. Code needs to be reviewed and depending on the outcome we'll need to decide what to do with it
C) Solution here depends on the former, do we require item_number on import or not? If yes then indeed we can always decide insert/update..

**Comment by jekkos** [comment]
There might be some usecases for items without item_numbers, which are eg service items, like hours of labor or generic items that represent products not counted in stock anymore. Currently all items have stock counter linked and that might change in the future as described in #200

**Comment by i92guboj** [comment]
I created a new issue for this, since we are deviating from the pic_id issue. One thing at a time!

**Comment by i92guboj** [comment]
@jekkos 

> @i92guboj For SQL changes you indeed need to modify tables.sql and add the particular statement to the upgrade script. Next one needs to run a grunt build, see wiki for instructions on that. This will regenerate the final scripts.

I can't find any upgrade script, and haven't found the instructions for grunt stuff either. I'll keep looking around, it must be somewhere, but if you have a link or two to share... :)

Thanks for all the guidance, by the way. It's much appreciated. :)

**Comment by i92guboj** [comment]
I had node.js installed, so I ran this:

`$ npm install -g grunt-cli
$ cd <osposdir>
$ grunt
Fatal error: Unable to find local grunt.
[...]
$ npm install
# lots of stuff happened here, I hope I didn't break anything :D
$ grunt
<complained about something called Bower, told me to use --force>
$ grunt --force
<ran without problems, it seems>`

I will have to check what this is for, but I really don't have the time for that now. I only checked the database.sql file, and I saw that the changes I med into tables.sql "propagated" correctly. So, all bright and clean, hehe.

Still, I have no idea what the "upgrade script" is. I will try to figure.

**Comment by jekkos** [comment]
Regarding grunt, I guess that output might be sufficient. It does need bower sources to do a full minification build. But for the sql it's not necessary. It's best to install bower through npm `npm install -g bower` and then run `bower install` in the project directory.

**Comment by jekkos** [comment]
The upgrade script in this case is the 3.0.2_to_3.1 file which might not have been created yet.

**Comment by i92guboj** [comment]
Oh, I had already created a 3.0.1_to_3.2.sql file. I will rename it. Is that enough or do I need to "register" that file somehow in an automation script of some kind?

**Comment by i92guboj** [comment]
It's taking a bit to polish all the rough edges, I have little time for this, but as of now, import works fine, and the regular mechanism to upload files one-by-one also works. But I have to take care of a few things and test thoroughly so we won't have big problems with this.

The main thing left to do, which I have decided to include in this same path since it's highly related, is to settle on some mechanism to regulate what to do when a given image exists.

So, right now, I have done this:

1. I have renamed pic_id to pic_filename
2. I changed the data type to varchar(255)
3. I have modified the code accordingly
4. I have patched some functions, extension is now self contained so you don't have to guess it by using glob(), I haven't get rid of glob() but got rid or *'s and the like
5. I have created 3.0.2_to_3.1.sql which will alter the column
6. I have tested all that, images appear and seem to work in the preview table, and in the item properties dialog, I can upload, erase, etc etc. Looking at the table, file names are correctly stored. So far, so good.

Now, I want to modify the file upload mechanism, so that files will retain their name. The rationale behind this is to be able to reuse photos. 

Right now, I can import using CSV, put my files in the right place, and they will be shown. Many items can share the same photo without any problem.

But if I upload photos one by one using the UI, they are still numbered as before, and multiple copies of the same file can be uploaded with different names, which is inefficient, as I see it.

My plan, after giving it some thought, is to provide the same UI so that the user can upload the file, and, if the file name matches one that exists, ask the user whether to re-use the image or upload it again. Either way the item will be assigned the same file name as pic. The only difference is that if you re-upload the file, if will be overwritten, else, the existing file will be used. As said, it needs some more thought and polishing, there are some loose ends to tie, but just wanted to say all this aloud in case you have some pointers about this.

Also, and this is not directly related, but... does pic_thumbs() serve any purpose? I grepped for it the whole tree and it's not called anywhere.... is it a zombie or a newborn which hasn't found its purpose in life yet? :lol:

**Comment by i92guboj** [comment]
To answer myself, pic_thumbs() is getting called each time a thumbnail is requested. I have no idea **how/from where**, though.

Probably my complete ignorance about CI methods and stuff. I have googled around but surely this can only be fixed by taking a CI quick course :tongue: 

I can't say I get all the logic behind this. Thumbnail related urls seem a bit obscure and unnecessarily complex to me, but I am not getting into that, at least for now. It might be some kind of relic, backwards compatible paths or something. Will have to check url rewrites, some day. That might shade some light into it.

**Comment by jekkos** [comment]
Yes indeed. Pic thumbs renders a thumbnail and sends it back with the appropriate content type. I have just quickly reviewed this part and found that there are potential security issues. 

The code here can pick up eg javascript file and send it back as application/javascript content type after which it will be rendered by the browser. This can result in some kind of xss attack or even worse, if a user is able to upload a polyglot jpeg with embedded PHP code (by bypassing the CI filter), then it can be executed by calling this function.

I think I'll need to do some thorough testing of this functionality to see whether it can be exploited or not.

**Comment by i92guboj** [comment]
I am getting there, however, while re-thinking about preserving integrity of already existing installations, I notice that modifying the existing data in pic_filename (formerly, pic_id), might not be that simple.

Since there's at least three extensions that people can use (jpg, gif and png), there's no easy way to guess what should go in there. 

OSPOS "guess" the extension by using glob(), which by the way I suspect isn't too secure either (I am no expert in PHP, came from C/C++, traditional stuff, you know ;) ). It just picks the first match, whatever that is, which can bring its own assortment of problems.

This part of the code has lots of these things, I have plans for that, but first I have to get familiar with all this CI chit-chat that I know nothing about :)

Getting back to the issue at hand, which would be the preferred way to deal with this? I could re-think the code so that, if it doesn't find the file name in the new way, it will try to guess the extension as before, but I'd rather not, because I'd prefer to ultimately get rid of all those .*'s in the code which I don't like at all. They seem to me like loose ends that need tying.

Some kind of shell script could do the job, but that won't be too cross-OS, I guess. I suppose there're people using this in non-Linux OSes, aren't there?

The original plan was to add something like this into the sql script, but that won't suffice, because of the many possible file extensions:

UPDATE `ospos_items` SET `pic_filename` = CONCAT(`pic_filename`, '.png');

**Comment by jekkos** [comment]
I guess we'd better solve this by updating those fields lazily in the PHP code, there's no other way correct to keep old installs working otherwise. Extensions could be uppercase, lowercase.. png, jpg, jpeg, you name it.

In this case I think it'd be good to have some kind of check that sees whether the extension is present in the database first. If it's not then the globbing function should just resolve it once more and update the field so that the next time it can use it directly.

After a while most of the fields will be updated and we'll be able to get rid of this legacy. 

In this case the thumbnail function could serve about anything as you say, setting the content type accordingly. Of course content types are filtered by CI on upload so we have to trust their implementation not to let malicious users bypass it.

**Comment by i92guboj** [comment]
Unfortunately, you are right. For now I am doing that. I don't like that, because this kind of stuff usually remains for ages in the source code. Once it gets there it probably won't go away for years. But, hell...

I've read in the past about this: 

http://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_load-file

However, it sends shivers down my spine to even think about that. Never used it, and certainly I can't think of a reliable way to determine the full path to the files on every server out there from the SQL upgrade script. Plus it's the kind of function that will work or fail depending on a gazillion FS-dependant issues.

**Comment by i92guboj** [comment]
I lazily update filenames in the table from Item->view(),  but I'd also like to update it from pic_thumb(), that will fix the tables much faster.

However, as said before, I have no idea where this func is called from, and I need to know that to change something. This little piece of code seems to get ext-less filenames, so I can't get rid of the globbing inside because of that.

Can you give me some pointer? I truly don't understand how it fits into scene. I know it's called each time you display the list of items, and, because of the way it acts, I also know it receives the extension-less filename as an argument, but I have no idea how. Grep tells me this function is never called from anywhere...

**Comment by i92guboj** [comment]
Neverming. It seems it's some arcane CI thing. I'll eventually get to that, but as of now, it suffices for me to know that this is what calls that func, somehow.

			$image .= '<a class="rollover" href="'. base_url($images[0]) .'"><img src="'.site_url('items/pic_thumb/' . pathinfo($item->pic_filename, PATHINFO_BASENAME)) . '"></a>';

I can work with that.

**Comment by jekkos** [comment]
Jep I guess that's the one calling this function. It uses the rollover class to activate a jQuery plugin that show the image as a thumbnail.

**Comment by i92guboj** [comment]
So I decided to fix it in search(), after calling get_item_data_row(), so that I don't have to reference the item from inside pic_thumb(). Right now, fixes go into the table when listing elements which I think is the best option.

At the current status of things, I need to get my lacky git-foo into shape to publish those changes. What do I need to create a PR so you can review the changes and, eventually, merge them if you think the patch is good enough?

Do I fork and work there or...? Note also that I know nothing at all about all this bower stuff and that. I usually just diff my .c(pp) and .h files and I never needed much else :P

I've read, once more, the wiki, but it doesn't tell much about development, other than configuring git to convert tabs to spaces, which I don't need because I already to that in my editor.

**Comment by jekkos** [comment]
Yes once you think you are done with these changes, you can create a pull request against the master branch of this repository. Creating Pull Requests is a github specific functionality and will compare two branches in different repositories and propose to merge the diff.

I see that you have the work in a branch in your repo, so you should then just select that branch as a source and the master of this repo as a target. Github will then create the PR and then we can have a further look at your changes and review them. Should be pretty straightforward.

If any change is required you just can keep working in that branch and the PR will be updated to include those

**Comment by i92guboj** [comment]
Hi, and thanks for the guidance.

I've used git in the past, just needed to refresh my skills, not much problem with that.

Knowing that, I will send the pr in the next few days. It could be later this evening or in a couple days, don't know when I will be able to do so. It shouldn't take long, though.

As for my github repo, I was just testing. I have not an exact idea on its current status.

I'll keep you informed. 

Thanks again :)

**Comment by i92guboj** [comment]
Just for future reference, the push request is in #1041

**Comment by jekkos** [comment]
Ok great, thanks, will look into as soon as I can.

**Comment by daN4cat** [comment]
@i92guboj it would be good to get this baselined so I can review it.

**Comment by daN4cat** [comment]
Code submitted to master.

**Comment by daN4cat** [comment]
More to test :-)

**Comment by daN4cat** [comment]
@i92guboj hey man, the code is quite buggy. Have you tested it? I keep fixing stuff....

**Comment by i92guboj** [comment]
Sorry, I see a few of these are leftovers of pic_id which I didn't notice. At some point, I grep'ed the sources but I either skipped those or messed up with git. Either way thanks for fixing. If I ever submit something else, I will also be more careful with indenting and such. I see you also fixed several of these. :(

**Comment by daN4cat** [comment]
@i92guboj can you please help with the testing and verify that what we have in master works properly?

**Comment by i92guboj** [comment]
Yes. I will do and report back.

Thanks!

**Comment by daN4cat** [comment]
@i92guboj I spotted an issue, here what happens:

- open an item with no image
- in the image box an icon "Avatar" is shown
- remove that image
- save the item
- the Items table is refreshed, all seems fine
- change view (e.g. Customers)
- go back to Items view
- a broken image icon is shown on the line of the above changed item

Please fix this issue.

**Comment by i92guboj** [comment]
Yes, I noticed this as well. There's some problem while saving. I am looking into this.

Thanks for confirming. I will try to get it fixed in the following hours.

**Comment by daN4cat** [comment]
Also the avatar icon in items with no image does not look right to me.
It should be empty as it used to be.

Thanks to look into this.

**Comment by i92guboj** [comment]
Yes. Something changed at some point and I don't know what it might be, I am trying to determine when this stopped working, because, at some point, it defintely worked with a hitch.

Unfortunattely there's something that's preventing me from being able to create or modify items, and dthe error message are completely useless when it comes to diagnosing problems, which in turn prevent's me from being able to check the issue at hand. Can't know if both things are linked or if there's some other issue with my installation so I will have to set this thing up from scratch, to discard some things.

**Comment by i92guboj** [comment]
Ok, I have narrowed down the thing a bit. It seems that, for items with empty pic_filename, saving the item again causes this field to be filled with a dot. Saving it again, adds an extra dot, so the value in the fields ends being "..", and it stops at this point (no further dots are added.

I have some ideas on what might be causing this. However, I also need to figure why simply opening the window to edit an item, save() seems to be triggered. I know it's called when listing the items (because I did that), but something else is triggering a save() when this window is openened.

In other words, if I open a item which has image, and delete the image, and save the item, at this point, pic_filename is NULL.

If I press the icon to edit the item again and check the database, this field contains ".", and I have no idea why.

If I sort that out, the patch should be simple enough.

Oh, the weird icon is part of the same problem. If the field is truly empty, as it should, the artifact doesn't show.

**Comment by i92guboj** [comment]
Please, review my new PR and see if it fixes the problems for you. If so, I can probably simplify the code and clean it up a bit. :)

**Comment by daN4cat** [comment]
@i92guboj thanks for the changes but please look into my review feedback before I can take your changes in master.

**Comment by i92guboj** [comment]
I pushed a couple commits. My tests are positive right now. Please let me know any issue.

Thank you for your patience! :)

**Comment by daN4cat** [comment]
@i92guboj almost there.... I tested but preexisting items with no image stored at Item edit time show the "Avatar" icon instead of nothing.
Still one extra step....

**Comment by daN4cat** [comment]
About the broken image, that's sorted. Thanks.

**Comment by i92guboj** [comment]
I am not sure I understand you. You mean that items that have the old pic_id contents in pic_filename are not displayed correctly?

What I intended in those cases is that file names would be automatically updated and images displayed.

If not, what's actually in those fields if you look in the db? Is it null, empty space, empty string or what?

**Comment by daN4cat** [comment]
Old items with no image show in the Edit modal window an "avatar" icon instead of being empty. If you remove that image and save the next time you edit it it's empty as you would expect on the first time.

**Comment by i92guboj** [comment]
But do you know what's in pic_filename in the ospos_items table when this happens? If you do a "select pic_filename from ospos_items where <that concrete product>" what do you get?

If it's empty or null, it shouldn't be different from the case when you empty the field yourself and then re-save the product...

So either there's some kind of corrupted contents in this field (either from before of from the conversion process) or there's some other problem in the code.

Thanks again for the feedback.

**Comment by daN4cat** [comment]
I'll have a look and will let you know.

**Comment by i92guboj** [comment]
I've been looking around and can't reproduce this, so any guidance on how to replicate that behavior is welcome. It seems to be working well, but it won't be until todays evening which I will put this to real use hehe, let's see what happens then.

:)

**Comment by daN4cat** [comment]
The pic_filename has two dots .. 
It could be the side effect of previous version that was not working properly and I saved the two dots.

I checked other items down the long list and they don't have this issue. So I would say it's a test envirnoment issue.

**Comment by i92guboj** [comment]
Yes, that's what I had in mind. 

Items without image (pic_filename=NULL), mused to mutate to single dot, and then to double dot in the bogus version of my patch. Note that those were updated simply when listing, you didn't really have to edit the items to get pic_filename modified since NULLs were in the "no-extension" code path.

What I think was happening here, is that glob was including current dir and parent dir in the file name guessing mechanism for this particular case when pic_filename was NULL. Add a dot as extension separator, let glob guess the file name, and, voilà! Bug ahead.

This shouldn't happen after that last commit but if needed we could put some extra checks to handle special filenames, such as dot and dot-dot, and empty pic_filename in these cases.

All in all, what you experienced here describes the same thing I saw while testing my last changes, and fits perfectly with my own experience, which is good since we have an explanation for that.

Thanks for taking the time to report back :)
